service/controller/resource/subnet/resource.go

```diff
@@ -215,6 +216,8 @@ func (r *Resource) ensureSubnets(ctx context.Context, deploymentsClient *azurere
 				return microerror.Mask(err)
 			}
 
+			allocatedSubnet.ID = subnetID
```

Comment:
Did you test if this works? I'd imagine `allocatedSubnet` to be copied instance of the element from `azureCluster` subnets. If that's the case, then the ID update here won't get effective when updating the `azureCluster` object. I'd maybe change the for loop to `for i := 0; i < len(azureCluster.Spec.NetworkSpec.Subnets); i++` form and then you could change the corresponding line here to:
`azureCluster.Spec.NetworkSpec.Subnets[i].ID = subnetID`